### PR TITLE
Add ability to build with memcheck framework and undefined behavior sanitizer

### PR DESCRIPTION
**Pull request by maxtmann**
*Description of changes:*

* Add undefined behavior sanitizer (UBSAN) and memcheck frameworks (ASAN and Valgrind) as configure options.
* Add fine-grained memory access checks and redzones to freelist (in case memory checks are enabled)

This PR also fixes some memory leaks detected by the memcheck frameworks. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -112,6 +112,36 @@ To enable UBSAN (Undefined Behaviour Sanitizer), use the following config option
...
 
+To enable memory access checks with ASAN (disabled by default), use
+the following config option:
+
+```
+   --enable-asan           Enable ASAN memory access checks
+```
+
+In case plugin is configured with `--enable-asan`, the user should run
+the application with `LD_PRELOAD=<path to libasan.so>` since plugin is
```
Comment (bwbarrett) [inline README.md:123]: I don't follow this comment; normal dynamic loading rules should work, no?

File: configure.ac [hunk 2]
```diff
@@ -121,6 +123,9 @@ AS_IF([test "${NCCL_OFI_PLATFORM}" = "none"], [AX_CHECK_PLATFORM_AWS()])
 
 AC_SUBST([NCCL_NET_OFI_DISTCHCK_CONFIGURE_FLAGS])
 
+AS_IF([test -n "${with_valgrind}" -a "${with_valgrind}" != "no" -a "${enable_asan}" = "yes"],
```
Comment (bwbarrett) [inline configure.ac:127]: This is not the exact same valgrind test you used in the valgrind configure.  Either add arguments to the valgrind macro to figure out if its enabled or not, or leak a variable that tells you if it is enabled or not.  But this is overly confusing for no reason.

File: include/nccl_ofi_memcheck_asan.h [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+#ifdef _cplusplus
+extern "C" {
+#endif
+
+#if !defined(__SANITIZE_ADDRESS__)
+#error "memcheck-asan should not be compiled when ASAN is disabled"
+#endif
+
+/*
+ * Marks memory region [addr, addr+size) as unaddressable.
```
Comment (bwbarrett) [inline include/nccl_ofi_memcheck_asan.h:17]: shouldn't you get these from `sanitizer/asan_interface.h`?

File: include/nccl_ofi_memcheck_builtin.h [hunk 4]
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2014-2023 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ */
+
+#ifndef NCCL_OFI_MEMCHECK_BUILTIN_H
```
Comment (bwbarrett) [inline include/nccl_ofi_memcheck_builtin.h:5]: why does this exist?  what does the builtin memcheck do?

File: src/nccl_ofi_freelist.c [hunk 5]
```diff
@@ -62,7 +62,7 @@ static int freelist_init_internal(size_t entry_size,
 		return -ENOMEM;
 	}
 
-	freelist->entry_size = NCCL_OFI_ROUND_UP(NCCL_OFI_MAX(entry_size, sizeof(struct nccl_ofi_freelist_elem_t)), 8);
+	freelist->entry_size = NCCL_OFI_ROUND_UP(NCCL_OFI_MAX(entry_size, sizeof(struct nccl_ofi_freelist_elem_t)), MEMCHECK_GRANULARITY);
```
Comment (bwbarrett) [inline src/nccl_ofi_freelist.c:71]: what guarantees that memcheck_granularity is not smaller than a pointer?

File: src/nccl_ofi_freelist.c [hunk 6]
```diff
@@ -153,6 +153,9 @@ int nccl_ofi_freelist_fini(nccl_ofi_freelist_t *freelist)
 
 	while (freelist->blocks) {
 		struct nccl_ofi_freelist_block_t *block = freelist->blocks;
+		nccl_net_ofi_mem_defined(block, sizeof(struct nccl_ofi_freelist_block_t));
+		void *memory = block->memory;
```
Comment (bwbarrett) [inline src/nccl_ofi_freelist.c:164]: this seems like an unnecessary change.

File: src/nccl_ofi_freelist.c [hunk 7]
```diff
@@ -265,5 +269,7 @@ int nccl_ofi_freelist_add(nccl_ofi_freelist_t *freelist,
 		buffer += freelist->entry_size;
 	}
 
+	nccl_net_ofi_mem_noaccess(block, sizeof(struct nccl_ofi_freelist_block_t));
```
Comment (bwbarrett) [inline src/nccl_ofi_freelist.c:290]: add a comment about why this is true.

File: include/nccl_ofi_freelist.h [hunk 8]
```diff
@@ -29,6 +29,9 @@ struct nccl_ofi_freelist_elem_t {
 	struct nccl_ofi_freelist_elem_t *next;
 };
 
+_Static_assert(sizeof(struct nccl_ofi_freelist_elem_t) % MEMCHECK_GRANULARITY == 0,
```
Comment (bwbarrett) [inline include/nccl_ofi_freelist.h:31]: Don't add an assert unless you have code to make it true.  Don't make the next person do the hard work.

File: tests/unit/freelist.c [hunk 9]
```diff
@@ -39,7 +39,7 @@ int deregmr_simple(void *handle)
 }
 
 struct random_freelisted_item {
-	int random;
+	size_t random;
```
Comment (bwbarrett) [inline tests/unit/freelist.c:42]: nack; the choice of `int` was intentional here, to make sure that we didn't have crazy alignment requirements on reginfo.

**Review by bwbarrett — CHANGES_REQUESTED** [review]
The changes to fix memory leaks should be their own PR.

The changes to add marking / unmarking of freelist should all be one commit.

The changes to support redzones should all be one commit.

File: README.md (hunk 1 above)
Comment (maxtmann) [reply README.md:123]: It does not seem like it will work. Without preload, we get
```
ASan runtime does not come first in initial library list; you should either link runtime to your application or manually preload it with LD_PRELOAD.
```
This approach is recommended in case app does not link ASAN and cannot be recompiled https://github.com/google/sanitizers/wiki/AddressSanitizerAsDso

Linking statically does not seem to work (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=64234), either.

File: include/nccl_ofi_memcheck_builtin.h (hunk 4 above)
Comment (maxtmann) [reply include/nccl_ofi_memcheck_builtin.h:5]: This header file is included in case no memory access checker is enabled. In this case it provides empty function definitions.

File: include/nccl_ofi_memcheck_asan.h (hunk 3 above)
Comment (maxtmann) [reply include/nccl_ofi_memcheck_asan.h:17]: I will remove the declarations and include the header in the next revision.

File: include/nccl_ofi_memcheck_builtin.h (hunk 4 above)
Comment (bwbarrett) [reply include/nccl_ofi_memcheck_builtin.h:5]: Sorry, that was a request for documentation.

File: configure.ac (hunk 2 above)
Comment (maxtmann) [reply configure.ac:127]: I would vote for the argument approach. I assume you mean that we pass the argument name as parameter and then access it via `$1` and `${$1}`. This boils down to a variable leakage as well I think but makes it more obvious in the `configure.ac` file.

File: include/nccl_ofi_memcheck_builtin.h (hunk 4 above)
Comment (maxtmann) [reply include/nccl_ofi_memcheck_builtin.h:5]: I see.

File: src/nccl_ofi_freelist.c (hunk 6 above)
Comment (maxtmann) [reply src/nccl_ofi_freelist.c:164]: In `nccl_ofi_freelist_add()`, we set the block structure to `noaccess`. In `nccl_ofi_freelist_fini()`, we want to read the member `memory` of that structure. Therefore, we need to set it back to defined, or do I miss something here?

File: configure.ac (hunk 2 above)
Comment (bwbarrett) [reply configure.ac:127]: Normally, you would have two arguments, the one evaluated if the package is found and the one that is evaluated if the package is not found.

File: README.md [hunk 10]
```diff
@@ -112,6 +112,36 @@ To enable UBSAN (Undefined Behaviour Sanitizer), use the following config option
...
+dynamically linked to libasan.so.
+
+In case plugin is configured with `--enable-asan` and the plugin is
```
[the remark this answers is not in the document]
Comment (maxtmann) [reply README.md:126]: I assume that a large amount users drives plugin with NCCL, and thus, with CUDA.

File: configure.ac (hunk 2 above)
Comment (bwbarrett) [reply configure.ac:127]: Autoconf style is that instead of:
```
CHECK_VALGRIND([valgrind_enabled])
```
and then having `CHECK_VALGRIND` set `valgrind_enabled` to 0 or 1, you instead do:
```
CHECK_VALGRIND([valgrind_enabled=1], [valgrind_enabled=0])
```
and the `CHECK_VALGRIND` macro has a line like this:
```
AS_IF([test "$my_happy_check" = "yes"], [$1], [$2])
```

Comment (maxtmann) [reply configure.ac:127]: Thank you.

File: src/nccl_ofi_freelist.c (hunk 6 above)
Comment (maxtmann) [reply src/nccl_ofi_freelist.c]: My take is that I need variables to temporarily store the pointer to the block memory and its size.

The reason is that the memcheck guards of the block memory are reset to undefined via call to `nccl_net_ofi_dealloc_mr_buffer()` before unmapping the memory such that the guards are correctly set when the same memory range is reallocated via mmap later. However, when I pass the pointer to the memory (size as well) directly to `nccl_net_ofi_dealloc_mr_buffer()` via `block->memory`, valgrind detects access to uninitialised value.

Comment (bwbarrett) [reply src/nccl_ofi_freelist.c]: yeah, I think the answer is that you don't need to re-add the guards there; you're about to free the memory, which has the same impact.

**Review by bwbarrett — APPROVED** [review]
